lib: Add AT shell library

jtguggedal opened the pull request:
Add library that enables AT command when shell is used.
Enabling the library will add a command called 'at' to the shell
which allows sending AT commands to the modem and receiving the
response back.
The command has two subcommands to enable and disable AT notifications
by enabling/disabling an AT command event handler.

jtguggedal commented:
@jhirsi @trantanen @tokangas I needed AT commands using shell in a project, and shamelessly based that integration on your work in the modem shell. It made me think that there is a use-case for easily enabling AT commands in shell in NCS, so extracted it as a library in this PR.
I know it's not quite what you need in mosh with the more sophisticated printing, but maybe we can accomodate also MOSH here? Maybe I could add a wrapper over the shell printing and implement it as a weak function, and then you can override it in the modem shell?

trantanen commented on lib/at_shell/CMakeLists.txt:
```diff
@@ -0,0 +1,10 @@
+# SPDX-License-Identifier: LicenseRef-Nordic-5-Clause
+#
+
+if(CONFIG_AT_SHELL)
```
Is this flagging needed? We have flags at higher level protecting the compilation of this library.

jtguggedal replied:
You're right, this is not needed, will remove

trantanen commented on lib/at_shell/at_shell.c:
```diff
@@ -0,0 +1,85 @@
+			return -EINVAL;
+		}
+
+		shell_print(shell, "%s", response);
```
As you mentioned in your comments, some application may want to use their own printing functions and not `shell_print()` directly to put some application specific formatting into all prints.
Giving a function pointer to use for printing would be a solution but there might still be issues with shell pointer as we avoid passing that in all mosh_print's in modem shell.

jtguggedal replied:
I'll see if I can come up with a suggestion, possibly in a follow-up PR

trantanen reviewed (commented):
This library is something that's been requested earlier when this functionality was made to modem shell so a good initiative to actually do it :)

jtguggedal commented on lib/at_shell/at_shell.c:
```diff
@@ -0,0 +1,85 @@
+
+static void at_notif_handler(const char *response)
+{
+	shell_print(global_shell, "AT event handler: \n%s", response);
```
I thought about it, and think we're okay here. The AT monitor is initialized in paused state, so it should not be possible to receive notifications before events have been enabled using the shell, and at that point `global_shell` is set to `shell` in the

trantanen replied:
Maybe an assert in that case?

tokangas commented on lib/at_shell/Kconfig:
```diff
@@ -0,0 +1,26 @@
+
+module = AT_SHELL
+module-str = AT shell
+source "${ZEPHYR_BASE}/subsys/logging/Kconfig.template.log_config"
```
Currently logging is not used, so this is probably not needed.

tokangas commented on lib/at_shell/at_shell.c:
```diff
@@ -0,0 +1,85 @@
+
+static void at_print_error_info(const struct shell *shell, int error)
+{
+	switch (nrf_modem_at_err_type(error)) {
+	case NRF_MODEM_AT_ERROR:
+		shell_print(shell, "ERROR");
+		break;
+	case NRF_MODEM_AT_CME_ERROR:
+		shell_print(shell, "CME ERROR: %d", nrf_modem_at_err(error));
+		break;
+	case NRF_MODEM_AT_CMS_ERROR:
+		shell_print(shell, "CMS ERROR: %d", nrf_modem_at_err(error));
+		break;
+	default:
+		break;
```
The default branch should never get executed, I guess, but this could be combined with NRF_MODEM_AT_ERROR, so it would at least print an error.
```suggestion
	switch (nrf_modem_at_err_type(error)) {
	case NRF_MODEM_AT_CME_ERROR:
		shell_print(shell, "CME ERROR: %d", nrf_modem_at_err(error));
		break;
	case NRF_MODEM_AT_CMS_ERROR:
		shell_print(shell, "CMS ERROR: %d", nrf_modem_at_err(error));
		break;
	case NRF_MODEM_AT_ERROR:
	default:
		shell_print(shell, "ERROR");
		break;
```

jtguggedal replied:
Good point, added as suggested

tokangas commented on lib/at_shell/at_shell.c:
```diff
@@ -0,0 +1,85 @@
+	int err;
+	static char response[CONFIG_AT_SHELL_CMD_RESPONSE_MAX_LEN + 1];
+
+	if (global_shell == NULL) {
+		global_shell = (struct shell *)shell;
+	}
```
This is fine as long as the `shell` pointer never changes. Just wondering if it can change or not? Anyway, removing the `if` would take care of that and the condition is kind of unnecessary anyway.

jtguggedal replied:
Hm, not sure if it can change, but I'll just remove the if anyway just in case.

NordicBuilder commented:
You can find the documentation preview for this PR at this <a href='http://developer.nordicsemi.com/nRF_Connect_SDK_dev/doc/PR-6094
'>link</a>. It will be updated about 10 minutes after the documentation build succeeds.

*Note: This comment is automatically posted by the Documentation Publishing GitHub Action.*

tokangas reviewed (approved):
Looks good to me, only a few comments.

rlubos reviewed (commented):
Looks pretty good, I think however this is such an interesting feature that it merits a changelog entry :)

trantanen commented on lib/at_shell/at_shell.c:
```diff
@@ -0,0 +1,85 @@
+	shell_print(global_shell, "AT event handler:\n%s", response);
+}
+
+static void at_print_error_info(const struct shell *shell, int error)
```
In `modem_shell`, I've done conversion from `at_cmd` library to `nrf_modem_at` few months ago and I did this similar function at first. Then I started to look at the response and it's actually pretty much the response coming from the modem so I ended up just printing modem response when err > 0. There are testing aspects why we could use `nrf_modem_at_err_type(error)` and `nrf_modem_at_err(error)` but other than that, the actual response from modem is really nice given it's formatted pretty much the way we want.

You can see MOSH PR on nrf_modem_at conversion now at:
https://github.com/nrfconnect/sdk-nrf/pull/6116

@tokangas, @jhirsi, Please also comment if you have a preference on which way to do it.

jhirsi replied:
Yes, modem response would be enough for me. Testing automation folks could be consulted as well.

I think that the additional feature that would add value would be to translate CME error values as text. There could one mapping api in NCS, although not sure if the at_shell is a correct place.

jtguggedal replied:
Good points. Seems like just printing the response is good enough here, so I'll remove `at_print_error_info` and just print it.
I saw some translation of CME codes somewhere, I'll see if I can dig them up. Can probably add that in another PR.

trantanen commented:
> Looks pretty good, I think however this is such an interesting feature that it merits a changelog entry :)

And don't we need library documentation?

trantanen commented on lib/at_shell/at_shell.c:
```diff
@@ -0,0 +1,85 @@
+
+static void at_notif_handler(const char *response)
+{
+	shell_print(global_shell, "AT event handler:\n%s", response);
```
As this is the only place where `global_shell` is used, maybe we could hide it's use into this function by using `shell_backend_uart_get_ptr()`.
Or, can we have some other than UART backed?

jtguggedal replied:
I think RTT backend would also be used by some

jtguggedal commented:
> Looks pretty good, I think however this is such an interesting feature that it merits a changelog entry :)

Added the addition of at_shell to changelog, and also added .rst for the library.

peknis commented on doc/nrf/libraries/modem/at_shell.rst:
```diff
@@ -0,0 +1,43 @@
+   :local:
+   :depth: 2
+
+The library can be used to add an AT shell command to a shell.
```
```suggestion
You can use this library to add an AT shell command to a shell.
```

peknis commented on doc/nrf/libraries/modem/at_shell.rst:
```diff
@@ -0,0 +1,43 @@
+This enables sending AT commands to the modem and receiving responses.
+When the library is enabled, the shell will have a command called ``at``.
+
+It's also possible to enable an AT command notifications.
```
```suggestion
You can also enable AT command notifications.
```

peknis commented on doc/nrf/libraries/modem/at_shell.rst:
```diff
@@ -0,0 +1,43 @@
+Configuration
+*************
+
+The library expects a shell to be enabled by the user.
```
```suggestion
The library expects that you have enabled a shell.
To enable and configure the shell, use the following Kconfig options:
```

peknis commented on doc/nrf/libraries/modem/at_shell.rst:
```diff
@@ -0,0 +1,43 @@
+* :kconfig:`CONFIG_AT_SHELL` - Enables the AT Host library
+* :kconfig:`CONFIG_AT_SHELL_CMD_RESPONSE_MAX_LEN` - Maximum AT command response length
+
+In order to send AT read commands (ending with ``?``), shell wildcards must be disabled:
```
```suggestion
To send AT read commands (ending with ``?``), disable the shell wildcards using the following Kconfig option:
```

peknis commented on doc/nrf/libraries/modem/at_shell.rst:
```diff
@@ -0,0 +1,43 @@
+   :depth: 2
+
+The library can be used to add an AT shell command to a shell.
+This enables sending AT commands to the modem and receiving responses.
```
```suggestion
This enables you to send AT commands to the modem and receive responses.
```

peknis commented on doc/nrf/libraries/modem/at_shell.rst:
```diff
@@ -0,0 +1,43 @@
+
+The library can be used to add an AT shell command to a shell.
+This enables sending AT commands to the modem and receiving responses.
+When the library is enabled, the shell will have a command called ``at``.
```
```suggestion
When the library is enabled, the shell will have a command ``at``.
```

peknis commented:
This library doc is not exactly aligned with the template (https://developer.nordicsemi.com/nRF_Connect_SDK/doc/latest/nrf/templates/library_template_README.html). Please check whether you would like to add something, for example the "Dependencies" section.

jtguggedal commented:
> This library doc is not exactly aligned with the template (https://developer.nordicsemi.com/nRF_Connect_SDK/doc/latest/nrf/templates/library_template_README.html). Please check whether you would like to add something, for example the "Dependencies" section.
> 
> It also looks like this needs a header file for generating the API documentation.

@peknis Thanks for the quick review! I've implemented your suggested changes and added a `Dependencies` section.

peknis commented on doc/nrf/libraries/modem/at_shell.rst:
```diff
@@ -0,0 +1,56 @@
+
+* :ref:`nrfxlib:nrf_modem_at`
+
+
```
```suggestion
API documentation
*****************

| Header file: :file:`path_to_the_header_file`
| Source file: :file:`lib/at_shell/at_shell.c`
```
Add the header file and the path to it.

jtguggedal replied:
The library actually doesn't have a header file and no public API, only Kconfig options, so I removed this section

trantanen replied:
Should we mention this, that is, the fact there is no header file. As a developer not familiar with shell and AT shell, I'd be looking for it. And not mentioning it in documentation can mean it doesn't exist, or it's been forgotten. So to rule this out. I'd clearly say somewhere that this library doesn't have an API or header file but only configurations.
I don't know if we've said this in some other library somewhere.

@peknis What do you think?

peknis replied:
A good idea. We have an example of such case in https://developer.nordicsemi.com/nRF_Connect_SDK/doc/latest/nrf/libraries/modem/at_host.html.

jtguggedal replied:
Alright, I'll add a similar section to that, thanks for the link @peknis 
Btw, there might be a header file coming down the road to allow for custom print/format functions

peknis commented on doc/nrf/libraries/modem/at_shell.rst:
```diff
@@ -0,0 +1,56 @@
+This enables you to send AT commands to the modem and receive responses.
+When the library is enabled, the shell will have a command ``at``.
+
+You can also enable AT command notifications.
```
```suggestion
You can also enable AT command notifications and disable them.
```

peknis commented on doc/nrf/libraries/modem/at_shell.rst:
```diff
@@ -0,0 +1,56 @@
+Dependencies
+************
+
+This library uses the following |NCS| libraries:
```
```suggestion
This library uses the following |NCS| library:
```

peknis commented on doc/nrf/libraries/modem/at_shell.rst:
```diff
@@ -0,0 +1,56 @@
+
+* :ref:`at_monitor_readme`
+
+This library uses the following `sdk-nrfxlib`_ libraries:
```
```suggestion
This library uses the following `sdk-nrfxlib`_ library:
```

simensrostad reviewed (approved):
Nice!

trantanen commented on doc/nrf/libraries/modem/at_shell.rst:
```diff
@@ -0,0 +1,50 @@
+*************
+
+The library expects that you have enabled a shell.
+To enable and configure the shell, use the following Kconfig options:
```
```suggestion
To enable and configure the AT shell, use the following Kconfig options:
```
Previous sentence said "enable a shell" and this sentence says "configure the shell" which refers to the general shell mentioned in previous sentence. At least for me. And this line is about AT shell, not shell in general.

jtguggedal replied:
Good point, I'll change as suggested

trantanen commented on doc/nrf/libraries/modem/at_shell.rst:
```diff
@@ -0,0 +1,50 @@
+
+* :kconfig:`CONFIG_AT_SHELL` - Enables the AT Host library
+* :kconfig:`CONFIG_AT_SHELL_CMD_RESPONSE_MAX_LEN` - Maximum AT command response length
+
```
Although it should be clear, should we mention that to enable the shell, `CONFIG_SHELL` should be set.

jtguggedal replied:
Will add a sentence about this

jtguggedal commented:
@trantanen @peknis I made some changes to the documentation after your suggestions, introducing a `Usage` section and moving text from the introduction into that section.

peknis commented on doc/nrf/libraries/modem/at_shell.rst:
```diff
@@ -0,0 +1,60 @@
+
+* :kconfig:`CONFIG_SHELL_WILDCARD`
+
+Dependencies
```
I would rather have the Usage section before the Dependencies.

jtguggedal replied:
Fixed
